packages/katello/katello/hostname-change.rb

```diff
@@ -239,6 +281,116 @@ def setup_opt_parser
+      end
+      new_vals.soa_admin_domain = "root.#{new_vals.zone}"
+      new_vals.key_file = @scenario_answers['foreman_proxy']['keyfile']
+      new_vals.old_fqdn = @old_hostname
+      new_vals.new_fqdn = @new_hostname
+
+      STDOUT.puts 'querying local DNS server...'
+      ip_serial_data = query_dns_server(@old_hostname, new_vals.zone, new_vals.local_ip)
+      new_vals.ip = ip_serial_data[:ip]
+      serial = ip_serial_data[:serial]
+      new_vals.new_serial = serial + 1 if serial
+      reverse_serial = query_dns_server(@old_hostname, new_vals.reverse_zone, new_vals.local_ip)[:serial]
+      new_vals.new_reverse_serial = reverse_serial + 1 if reverse_serial
+      run_cmd('rndc thaw')
+
+      new_vals
+    end
+
+    def nsupdate_command(dns_entries, key_file)
+      "echo -e \"#{dns_entries}\" | nsupdate -l -k #{key_file}"
+    end
+
+    def update_dns_records(d)
+      %w[local_ip zone reverse_zone soa_admin_domain key_file old_fqdn new_fqdn ip new_serial new_reverse_serial].each do |value|
+        unless d[value]
+          fail_with_message """
+  Error gathering DNS data: couldn't find value for '#{value}'.
+  Make sure /etc/foreman-installer/scenarios.d/#{@options[:scenario]}-answers.yaml
+  'foreman-proxy' section has values for dns_zone, dns_reverse, and keyfile.
+  Make sure A and SOA records are present for #{@old_hostname}.
+"""
+        end
+      end
+
+      STDOUT.puts 'updating DNS records:'
+      # Use nsupdate to update DNS records
+      # Update SOA record to new hostname; add new A and NS records; delete old A and NS records.
+      # Multi-line strings are not indented because Ruby 2.0 doesn't support <<~ heredocs
+
+      forward_dns_command = <<-HEREDOC
+local #{d.local_ip}
+zone #{d.zone}
+update add #{d.zone} 10800 SOA #{d.new_fqdn} #{d.soa_admin_domain}. #{d.new_serial} 86400 3600 604800 3600
+update add #{d.zone}. 3600 IN NS #{d.new_fqdn}.
+update delete #{d.zone}. IN NS #{d.old_fqdn}
+update delete #{d.old_fqdn} A
+update add #{d.new_fqdn} 86400 A #{d.ip}
+send
+      HEREDOC
+
+      reverse_dns_command = <<-HEREDOC
+local #{d.local_ip}
+zone #{d.reverse_zone}
+update add #{d.reverse_zone} 10800 SOA #{d.new_fqdn} root.#{d.reverse_zone}. #{d.new_reverse_serial} 86400 3600 604800 3600
+update add #{d.reverse_zone} 3600 IN NS #{d.new_fqdn}
+update delete #{d.reverse_zone} IN NS #{d.old_fqdn}
+send
+      HEREDOC
+
+      STDOUT.puts 'forward...'
+      run_cmd nsupdate_command(forward_dns_command, d.key_file)
+      STDOUT.puts 'reverse...'
+      run_cmd nsupdate_command(reverse_dns_command, d.key_file)
+      STDOUT.puts 'updating dynamic zone files...'
+      run_cmd('rndc freeze')
+      run_cmd('rndc thaw')
+      STDOUT.puts 'DNS records updated'
+    end
+
+    def restore_last_scenario_yaml
+      STDOUT.puts 'restoring last_scenario.yaml'
+      if File.exist?(last_scenario_yaml)
+        run_cmd("cp #{@temp_last_scenario_yaml.path} #{last_scenario_yaml}")
```

Comment:
Not sure what the convention is, but should this use native ruby methods to copy a file instead?

Reply:
eh, let's leave well-enough alone. it was already written this way, but I do agree on the change. this PR is big enough IMO